### DISCUSSION TITLE
How to visualize the flow of a 3D image registration?

### DISCUSSION THREAD

**Discussion by Uemerson** [voxelmorph/voxelmorph#580] [Q&A]
After registration, the network returns two values, the first being the registered image and the second being the flow.

In a 3D registration, the flow (in my case) has a shape of (3, 160, 192, 224).

I followed the tutorial at [here](https://colab.research.google.com/drive/1F8f1imh5WfyBv-crllfeJBFY16-KHl9c?usp=sharing#scrollTo=PT0Cy-UAGS_t) to plot, and the grid seems correct. However, when I plot the flow, it appears strange. Below is the code I used:

```
import matplotlib.pyplot as plt
import voxelmorph as vxm
import neurite as ne
import numpy as np
import pystrum
import scipy

volshape = [192, 224]
vol_shape = (160, 192, 224)

grid = pystrum.pynd.ndutils.bw_grid(vol_shape=volshape, spacing=11)
# ne.plot.slices(grid, width=3)

warp = disp_data['arr_0'].reshape(160, 192, 224, 3)[vol_shape[0]//2, :, :, 0:2]
# get a transfer model from voxelmorph and warp
trf = vxm.networks.Transform(volshape)
warped_grid = trf([grid[None, ..., None], warp[None]])

titles = ['grid', 'warped grid']
slices = [grid, warped_grid[0, ..., 0]]
ne.plot.slices(slices, width=6.5, titles=titles);

ne.plot.flow([warp], width=5);
```

Any suggestions on how to make it work, or if the above code is correct?

Results:

![image](https://github.com/voxelmorph/voxelmorph/assets/19247745/ae1a7e7d-156a-4746-a783-a3600d3d5f65)

![image](https://github.com/voxelmorph/voxelmorph/assets/19247745/77142423-57aa-49db-b7d7-77dac316f9ab)

**Comment by Uemerson** [comment]
if I change the code:

```
warp = disp_data['arr_0'].reshape(160, 192, 224, 3)[vol_shape[0]//2, :, :, 0:2]
```

to:

```
warp = disp_data['arr_0'].squeeze()[0:2, 160//2, ::, ::].reshape(192, 224, 2)
```

the results:

![image](https://github.com/voxelmorph/voxelmorph/assets/19247745/57878321-a526-4566-b1fb-c75da0696fad)

![image](https://github.com/voxelmorph/voxelmorph/assets/19247745/a05c4aaf-018c-495e-8c7d-d6fd24ed3f86)

**Reply by Uemerson** [reply]
@adalca can u help-me again ? :)

**Reply by adalca** [reply]
just to be clear, your original field is `[3, 160, 192, 224]` (I assume from torch?) and you're doing a reshape `disp_data['arr_0'].reshape(160, 192, 224, 3)` ?

Also what would you *expect* it to look like?

**Reply by Uemerson** [reply]
@adalca 

Yes, that's correct. After performing the 3D registration using PyTorch, I obtained the flow (in addition to the registered image) with a shape of [1, 3, 160, 192, 224].

What I need to do is to plot the flow like these two images:

![image](https://github.com/voxelmorph/voxelmorph/assets/19247745/1b20b1e1-01ab-443c-a1d0-5b5070a50b07)

How would I do this with 3D registration?
